linkerd/app/inbound/src/server.rs

```diff
@@ -16,72 +16,64 @@ struct TcpEndpoint {
 // === impl Inbound ===
 
 impl Inbound<()> {
-    pub fn serve<B, G, GSvc, P>(
+    pub async fn serve<L, A, I, G, GSvc, P>(
         self,
-        bind: B,
+        addr: Local<ServerAddr>,
+        listen: L,
         profiles: P,
         gateway: G,
-    ) -> (Local<ServerAddr>, impl Future<Output = ()> + Send)
-    where
-        B: Bind<ServerConfig>,
-        B::Addrs: svc::Param<Remote<ClientAddr>>
-            + svc::Param<Local<ServerAddr>>
-            + svc::Param<OrigDstAddr>,
+    ) where
+        L: Stream<Item = io::Result<(A, I)>> + Send + Sync + 'static,
```

Comment:
nit/TIOLI: seems like `L` could just be an `impl Trait`, like
```rust
    pub async fn serve<A, I, G, GSvc, P>(
        self,
        addr: Local<ServerAddr>,
        listen: impl Stream<Item = io::Result<(A, I)>> + Send + Sync + 'static,
        profiles: P,
        gateway: G,
    ) where
```
to avoid adding another type param?

not a big deal either way